main_test.go

```diff
@@ -1,20 +1,134 @@
+	createdProject := &Project{Name: "Project_A"}
+	DB.Create(&createdProject)
+	pid = createdProject.ID
 
-	DB.Create(&user)
+	// 2. create a workflow belonging to a project
+	createdWorkflow := &Workflow{Name: "Workflow_A", ProjectID: createdProject.ID}
+	DB.Create(&createdWorkflow)
+	wid = createdWorkflow.ID
 
-	var result User
-	if err := DB.First(&result, user.ID).Error; err != nil {
-		t.Errorf("Failed, got error: %v", err)
+	// 3. create a task belonging to a workflow
+	createdTask := &Task{Name: "Task_A", WorkflowID: createdWorkflow.ID}
+	DB.Create(&createdTask)
+	tid = createdTask.ID
+
+	// 4. create a task with prev deps
+	createdTaskWithPrevDeps := &Task{Name: "Task_B", WorkflowID: createdWorkflow.ID}
+	createdTaskWithPrevDeps.PrevDeps = append(createdTaskWithPrevDeps.PrevDeps, &Task{
+		Model: gorm.Model{ID: createdTask.ID},
+	})
+	DB.Create(&createdTaskWithPrevDeps)
+	nid = createdTaskWithPrevDeps.ID
+
+	if pid == 0 || wid == 0 || tid == 0 || nid == 0 {
+		return 0, 0, 0, 0, fmt.Errorf("failed to prepare data")
 	}
+
+	return pid, wid, tid, nid, nil
+}
+
+func TestDeleteProjectWithWorkflowTask(t *testing.T) {
+	pid, wid, tid, nid, err := prepare()
+	_ = assert.Nil(t, err)
+
+	var p Project
+	err = DB.First(&p, pid).Error
+	_ = assert.Nil(t, err)
+
+	// Check Point: if Project_A has been deleted, Workflow_A and Task_A should be deleted too
+	err = DB.Select("Workflows", "Workflows.Tasks", "Workflows.Tasks.NextDeps").Delete(&p).Error
+	_ = assert.Nil(t, err)
+
+	err = DB.First(&Project{}, pid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Workflow{}, wid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Task{}, tid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Task{}, nid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+}
+
+func TestDeleteWorkflowWithTask(t *testing.T) {
+	_, wid, tid, nid, err := prepare()
+	_ = assert.Nil(t, err)
+
+	var w Workflow
+	err = DB.First(&w, wid).Error
+	_ = assert.Nil(t, err)
+
+	// Check Point: if Workflow_A has been deleted, Task_A and Task_B should be deleted too
+	err = DB.Select("Tasks", "Tasks.NextDeps").Delete(&w).Error
+	_ = assert.Nil(t, err)
+
+	err = DB.First(&Workflow{}, wid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Task{}, tid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Task{}, nid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+}
+
+func TestDeleteTasksWithNextDeps(t *testing.T) {
+	_, _, tid, nid, err := prepare()
+	_ = assert.Nil(t, err)
+
+	var task Task
+	err = DB.First(&task, tid).Error
+	_ = assert.Nil(t, err)
+
+	// Check Point: if Task_A has been deleted, Task_B should be deleted too
+	err = DB.Select("NextDeps").Delete(&task).Error
+	_ = assert.Nil(t, err)
+
+	err = DB.First(&Task{}, tid).Error
+	_ = assert.Equal(t, gorm.ErrRecordNotFound, err)
+
+	err = DB.First(&Task{}, nid).Error
```

Comment:
This is not correct because you have a many-to-many relationship. For example, you can see that it is also working for Workflow and deleting the Workflow records, but for NextDeps, it is not deleting since it is a many-to-many relationship.

Reply:
> This is not correct because you have a many-to-many relationship. For example, you can see that it is working for Workflow and deleting the Workflow record, but for NextDeps, it is not deleting since it is a many-to-many relationship.

Two Question: 
1. > [ FAILED ] A has many B, and B has many C: deleting A will only delete B, it cannot delete C

2. why can not delete many-to-many relationships? I think it is a actual requirement, which should be implemented. 

In my scenario, 
- a task with PrevDeps and NextDeps
- if this task is deleted, I think its NextDeps should also be deleted.(No need for PrevDeps)

Reply:
The reason is for example you have many-to-many relation between a user and a role entity, so the user has many roles and roles has many users, now if for each user deletion we delete the roles as well it might also remove the roles of other users because roles can be assigned to many user.

Reply:
Yes, If the case you mentioned, your reply is right.
But, My scenario is `PrevTasks`->`Task`->`NextTasks`, 
If deleting `Task`, deleting the `NextTasks` by cascade is valid. Of course, here `PrevTasks` shouldn't be deleted.

Reply:
Here is a [Java Spring JPA Demo](https://github.com/pplmx/orm-demo), which supports the above cases. FYI.

Reply:
Please feel free to submit a feature PR or create a feature request issue.